File: backend/services/burnout_predictor.py
"""
Burnout Predictor Service
ML-inspired burnout prediction and risk assessment
"""

import math
from enum import Enum


class BurnoutRiskLevel(Enum):
    LOW = ("Low", "#10B981", 0.3)
    MODERATE = ("Moderate", "#F59E0B", 0.6)
    HIGH = ("High", "#EF4444", 0.8)
    CRITICAL = ("Critical", "#DC2626", 1.0)


class BurnoutPredictorService:
    """Predicts burnout risk using ML-inspired algorithms"""

    def __init__(self):
        self.session_history = []
        self.max_sessions = 50
        self.risk_factors = {}

    def calculate_blink_trend(self, blink_history):
        """Calculate blink rate decline risk"""
        if len(blink_history) < 5:
            return 0

        recent = blink_history[-5:]
        older = blink_history[-10:-5] if len(blink_history) >= 10 else recent

        recent_avg = sum(recent) / len(recent)
        older_avg = sum(older) / len(older) if older else recent_avg

        decline = max(0, older_avg - recent_avg)
        return min(100, (decline / max(older_avg, 1)) * 100)

    def calculate_yawn_risk(self, yawn_data):
        """Calculate yawn frequency risk"""
        total_yawns = yawn_data.get('totalYawns', 0)
        time_elapsed = max(1, yawn_data.get('timeElapsedMinutes', 1))

        yawns_per_minute = total_yawns / time_elapsed
        return min(100, yawns_per_minute * 40)

    def calculate_focus_degradation(self, focus_history):
        """Calculate focus decline over session"""
        if len(focus_history) < 10:
            return 0

        window_size = max(1, len(focus_history) // 3)
        first = focus_history[:window_size]
        last = focus_history[-window_size:]

        first_avg = sum(first) / len(first)
        last_avg = sum(last) / len(last)

        degradation = max(0, first_avg - last_avg)
        return min(100, (degradation / 100) * 100)

    def calculate_posture_decline(self, posture_history):
        """Calculate posture decline"""
        if len(posture_history) < 10:
            return 0

        window_size = max(1, len(posture_history) // 3)
        first = posture_history[:window_size]
        last = posture_history[-window_size:]

        first_avg = sum(first) / len(first)
        last_avg = sum(last) / len(last)

        decline = max(0, first_avg - last_avg)
        return min(100, (decline / 100) * 100)

    def calculate_fatigue_escalation(self, fatigue_history):
        """Calculate fatigue escalation"""
        if len(fatigue_history) < 5:
            return 0

        recent = fatigue_history[-5:]
        older = fatigue_history[-10:-5] if len(fatigue_history) >= 10 else recent

        recent_avg = sum(recent) / len(recent)
        older_avg = sum(older) / len(older) if older else recent_avg

        escalation = max(0, recent_avg - older_avg)
        return min(100, escalation)

    def predict_burnout_risk(self, current_metrics, historical_data):
        """Predict burnout risk (0-100)"""
        weights = {
            'blink_trend': 0.2,
            'yawn_frequency': 0.15,
            'focus_degradation': 0.25,
            'posture_decline': 0.15,
            'session_duration': 0.1,
            'fatigue_escalation': 0.15,
        }

        self.risk_factors = {
            'blink_trend': self.calculate_blink_trend(
                historical_data.get('blinkRate', [])
            ),
            'yawn_frequency': self.calculate_yawn_risk(
                current_metrics.get('yawnData', {})
            ),
            'focus_degradation': self.calculate_focus_degradation(
                historical_data.get('focus', [])
            ),
            'posture_decline': self.calculate_posture_decline(
                historical_data.get('posture', [])
            ),
            'session_duration': min(100, (current_metrics.get('sessionMinutes', 0) / 2)),
            'fatigue_escalation': self.calculate_fatigue_escalation(
                historical_data.get('fatigue', [])
            ),
        }

        risk_score = 0
        for factor, weight in weights.items():
            risk_score += self.risk_factors[factor] * weight

        return min(100, max(0, risk_score))

    def get_risk_level(self, risk_score):
        """Get burnout risk level"""
        if risk_score < BurnoutRiskLevel.MODERATE.value[2] * 100:
            return {
                'label': BurnoutRiskLevel.LOW.value[0],
                'color': BurnoutRiskLevel.LOW.value[1],
            }
        elif risk_score < BurnoutRiskLevel.HIGH.value[2] * 100:
            return {
                'label': BurnoutRiskLevel.MODERATE.value[0],
                'color': BurnoutRiskLevel.MODERATE.value[1],
            }
        elif risk_score < BurnoutRiskLevel.CRITICAL.value[2] * 100:
            return {
                'label': BurnoutRiskLevel.HIGH.value[0],
                'color': BurnoutRiskLevel.HIGH.value[1],
            }
        else:
            return {
                'label': BurnoutRiskLevel.CRITICAL.value[0],
                'color': BurnoutRiskLevel.CRITICAL.value[1],
            }

    def predict_burnout_timeline(self, risk_score, degradation_rate=0.5):
        """Predict minutes until burnout"""
        if risk_score < 50:
            return None

        minutes_to_critical = max(1, (100 - risk_score) / degradation_rate)
        return min(60, max(2, int(round(minutes_to_critical))))

    def get_recommendations(self, risk_score, risk_factors):
        """Get personalized recommendations"""
        recommendations = []

        if risk_factors.get('blink_trend', 0) > 40:
            recommendations.append({
                'priority': 'high',
                'action': 'Take a 20-20-20 eye break',
                'reason': 'Your blink rate is declining significantly',
                'duration': 1,
            })

        if risk_factors.get('yawn_frequency', 0) > 50:
            recommendations.append({
                'priority': 'critical',
                'action': 'Take a 5-10 minute rest',
                'reason': 'High frequency of yawns detected',
                'duration': 5,
            })

        if risk_factors.get('focus_degradation', 0) > 30:
            recommendations.append({
                'priority': 'medium',
                'action': 'Try a quick meditation or breathing exercise',
                'reason': 'Your focus is gradually declining',
                'duration': 3,
            })

        if risk_factors.get('posture_decline', 0) > 30:
            recommendations.append({
                'priority': 'medium',
                'action': 'Perform posture-correcting exercises',
                'reason': 'Your posture has significantly declined',
                'duration': 5,
            })

        if risk_factors.get('session_duration', 0) > 70:
            recommendations.append({
                'priority': 'high',
                'action': 'Take a longer break or end session',
                'reason': 'You\'ve been working for an extended period',
                'duration': 15,
            })

        if risk_score > 70:
            recommendations.append({
                'priority': 'critical',
                'action': 'Take an extended break NOW',
                'reason': 'Burnout risk is critically high',
                'duration': 30,
            })

        return recommendations


def create_burnout_predictor():
    """Factory function"""
    return BurnoutPredictorService()

File: frontend/src/utils/wellness/burnoutPredictor.js
// Predictive Burnout Detection Engine
// ML-inspired algorithms to predict burnout before it happens

import { BURNOUT_RISK_LEVELS, PREDICTION_WINDOWS, TREND_PERIODS } from '../../constants/wellnessThresholds.js';

class BurnoutPredictor {
  constructor() {
    this.sessionMetrics = [];
    this.maxSessions = 50;
    this.riskFactors = {
      blinkTrend: 0,
      yawnFrequency: 0,
      focusDegradation: 0,
      postureDecline: 0,
      sessionDuration: 0,
      fatigueEscalation: 0,
    };
  }

  /**
   * Calculate blink rate trend (declining blink = higher risk)
   */
  calculateBlinkTrend(blinkHistory) {
    if (blinkHistory.length < 5) return 0;

    const recent = blinkHistory.slice(-5);
    const older = blinkHistory.slice(-10, -5);

    const recentAvg = recent.reduce((a, b) => a + b, 0) / recent.length;
    const olderAvg = older.reduce((a, b) => a + b, 0) / older.length;

    const decline = Math.max(0, olderAvg - recentAvg);
    return Math.min(100, (decline / olderAvg) * 100);
  }

  /**
   * Calculate yawn frequency risk
   * Higher frequency = higher burnout risk
   */
  calculateYawnRisk(yawnData) {
    const { totalYawns = 0, timeElapsedMinutes = 1 } = yawnData;
    const yawnsPerMinute = totalYawns / Math.max(1, timeElapsedMinutes);

    // Normal: 0-1 yawn per minute
    // Risk: 1-2 yawns per minute
    // High risk: >2 yawns per minute
    return Math.min(100, yawnsPerMinute * 40);
  }

  /**
   * Calculate focus degradation over session
   */
  calculateFocusDegradation(focusHistory) {
    if (focusHistory.length < 10) return 0;

    const windowSize = Math.floor(focusHistory.length / 3);
    const first = focusHistory.slice(0, windowSize);
    const last = focusHistory.slice(-windowSize);

    const firstAvg = first.reduce((a, b) => a + b, 0) / first.length;
    const lastAvg = last.reduce((a, b) => a + b, 0) / last.length;

    const degradation = Math.max(0, firstAvg - lastAvg);
    return Math.min(100, (degradation / 100) * 100);
  }

  /**
   * Calculate posture decline risk
   */
  calculatePostureDecline(postureHistory) {
    if (postureHistory.length < 10) return 0;

    const windowSize = Math.floor(postureHistory.length / 3);
    const first = postureHistory.slice(0, windowSize);
    const last = postureHistory.slice(-windowSize);

    const firstAvg = first.reduce((a, b) => a + b, 0) / first.length;
    const lastAvg = last.reduce((a, b) => a + b, 0) / last.length;

    const decline = Math.max(0, firstAvg - lastAvg);
    return Math.min(100, (decline / 100) * 100);
  }

  /**
   * Calculate fatigue escalation risk
   */
  calculateFatigueEscalation(fatigueHistory) {
    if (fatigueHistory.length < 5) return 0;

    const recent = fatigueHistory.slice(-5);
    const older = fatigueHistory.slice(-10, -5);

    const recentAvg = recent.reduce((a, b) => a + b, 0) / recent.length;
    const olderAvg = older.reduce((a, b) => a + b, 0) / older.length;

    const escalation = Math.max(0, recentAvg - olderAvg);
    return Math.min(100, escalation);
  }

  /**
   * Predict burnout risk (0-100)
   */
  predictBurnoutRisk(currentMetrics, historicalData) {
    const weights = {
      blinkTrend: 0.2,
      yawnFrequency: 0.15,
      focusDegradation: 0.25,
      postureDecline: 0.15,
      sessionDuration: 0.1,
      fatigueEscalation: 0.15,
    };

    this.riskFactors = {
      blinkTrend: this.calculateBlinkTrend(historicalData.blinkRate || []),
      yawnFrequency: this.calculateYawnRisk(currentMetrics.yawnData || {}),
      focusDegradation: this.calculateFocusDegradation(historicalData.focus || []),
      postureDecline: this.calculatePostureDecline(historicalData.posture || []),
      sessionDuration: Math.min(100, (currentMetrics.sessionMinutes || 0) / 2),
      fatigueEscalation: this.calculateFatigueEscalation(historicalData.fatigue || []),
    };

    let riskScore = 0;
    for (const [factor, weight] of Object.entries(weights)) {
      riskScore += this.riskFactors[factor] * weight;
    }

    return Math.min(100, Math.max(0, riskScore));
  }

  /**
   * Get burnout risk level
   */
  getRiskLevel(riskScore) {
    if (riskScore < BURNOUT_RISK_LEVELS.MODERATE.threshold * 100)
      return BURNOUT_RISK_LEVELS.LOW;
    if (riskScore < BURNOUT_RISK_LEVELS.HIGH.threshold * 100)
      return BURNOUT_RISK_LEVELS.MODERATE;
    if (riskScore < BURNOUT_RISK_LEVELS.CRITICAL.threshold * 100)
      return BURNOUT_RISK_LEVELS.HIGH;
    return BURNOUT_RISK_LEVELS.CRITICAL;
  }

  /**
   * Predict burnout timeline
   * Returns minutes until predicted burnout
   */
  predictBurnoutTimeline(riskScore, degradationRate) {
    if (riskScore < 50) return null; // Low risk, no prediction

    const riskPerMinute = degradationRate || 0.5;
    const minutesToCritical = Math.max(1, (100 - riskScore) / riskPerMinute);

    return Math.min(60, Math.max(2, Math.round(minutesToCritical)));
  }

  /**
   * Get personalized recommendations based on risk
   */
  getRecommendations(riskScore, riskFactors) {
    const recommendations = [];

    if (riskFactors.blinkTrend > 40) {
      recommendations.push({
        priority: 'high',
        action: 'Take a 20-20-20 eye break',
        reason: 'Your blink rate is declining significantly',
        duration: 1,
      });
    }

    if (riskFactors.yawnFrequency > 50) {
      recommendations.push({
        priority: 'critical',
        action: 'Take a 5-10 minute rest',
        reason: 'High frequency of yawns detected',
        duration: 5,
      });
    }

    if (riskFactors.focusDegradation > 30) {
      recommendations.push({
        priority: 'medium',
        action: 'Try a quick meditation or breathing exercise',
        reason: 'Your focus is gradually declining',
        duration: 3,
      });
    }

    if (riskFactors.postureDecline > 30) {
      recommendations.push({
        priority: 'medium',
        action: 'Perform posture-correcting exercises',
        reason: 'Your posture has significantly declined',
        duration: 5,
      });
    }

    if (riskFactors.sessionDuration > 70) {
      recommendations.push({
        priority: 'high',
        action: 'Take a longer break or end session',
        reason: "You've been working for an extended period",
        duration: 15,
      });
    }

    if (riskScore > 70) {
      recommendations.push({
        priority: 'critical',
        action: 'Take an extended break NOW',
        reason: 'Burnout risk is critically high',
        duration: 30,
      });
    }

    return recommendations;
  }

  /**
   * Store session metrics
   */
  recordSession(sessionData) {
    this.sessionMetrics.push({
      timestamp: Date.now(),
      ...sessionData,
    });

    if (this.sessionMetrics.length > this.maxSessions) {
      this.sessionMetrics.shift();
    }
  }

  /**
   * Get historical trends
   */
  getHistoricalTrends() {
    if (this.sessionMetrics.length < 2) return null;

    const recent = this.sessionMetrics.slice(-5);
    const avgRiskScore = recent.reduce((a, b) => a + (b.riskScore || 0), 0) / recent.length;
    const trend =
      recent[recent.length - 1].riskScore > recent[0].riskScore ? 'worsening' : 'improving';

    return {
      avgRiskScore,
      trend,
      sessionsTracked: this.sessionMetrics.length,
    };
  }

  /**
   * Reset for new session
   */
  resetSession() {
    this.riskFactors = {
      blinkTrend: 0,
      yawnFrequency: 0,
      focusDegradation: 0,
      postureDecline: 0,
      sessionDuration: 0,
      fatigueEscalation: 0,
    };
  }
}

const burnoutPredictor = new BurnoutPredictor();

export const predictBurnout = (scoreHistory, sessionData) => {
  if (!scoreHistory || scoreHistory.length === 0) {
    return {
      riskScore: 0,
      riskLevel: 'low',
      recommendations: [],
      timeline: null,
    };
  }

  const currentMetrics = scoreHistory[scoreHistory.length - 1] || {};
  const historicalData = {
    posture: scoreHistory.map(s => s.posture || 0),
    eyeHealth: scoreHistory.map(s => s.eyeHealth || 0),
    focus: scoreHistory.map(s => s.focus || 0),
    fatigue: scoreHistory.map(s => s.fatigue || 0),
  };

  const riskScore = burnoutPredictor.predictBurnoutRisk(currentMetrics, historicalData);
  const riskLevel = burnoutPredictor.getRiskLevel(riskScore);
  const timeline = burnoutPredictor.predictBurnoutTimeline(
    riskScore,
    (scoreHistory.length > 1 ? Math.abs(scoreHistory[scoreHistory.length - 1].wellnessIndex - scoreHistory[scoreHistory.length - 2].wellnessIndex) : 0) / 10
  );
  const recommendations = burnoutPredictor.getRecommendations(riskScore, burnoutPredictor.riskFactors);

  return {
    riskScore,
    riskLevel,
    recommendations,
    timeline,
  };
};

export default burnoutPredictor;
